File: problema 19/fortran/problema_19.f90
module matrixOps
    implicit none
    contains

    subroutine printMatrix(matrix)
        real, intent(in) :: matrix(:,:)
        integer :: i, j
        do i = 1, size(matrix, 1)
            do j = 1, size(matrix, 2)
                write(*, '(F6.1)', advance='no') matrix(i, j)
            end do
            write(*,*)  ! Ensure new line after each row
        end do
    end subroutine printMatrix

    subroutine reverseRows(matrix)
        real, intent(inout) :: matrix(:,:)
        integer :: i, j, m, n
        real :: temp
        m = size(matrix, 1)
        n = size(matrix, 2)

        do i = 1, m / 2
            do j = 1, n
                temp = matrix(i, j)
                matrix(i, j) = matrix(m - i + 1, j)
                matrix(m - i + 1, j) = temp
            end do
        end do
    end subroutine reverseRows

    subroutine modifyRows(matrix)
        real, intent(inout) :: matrix(:,:)
        integer :: i, j, m, n
        real, allocatable :: tempRow(:)
        m = size(matrix, 1)
        n = size(matrix, 2)

        allocate(tempRow(n))

        do i = 1, m
            tempRow = matrix(i, :) * m + (i - 1) * matrix(m, :)
            matrix(i, :) = tempRow
        end do

        deallocate(tempRow)
    end subroutine modifyRows

end module matrixOps

program matrix_operations
    use matrixOps
    implicit none
    integer :: m, n, i, j
    real, allocatable :: matrix(:,:)

    ! Read matrix dimensions
    print *, "Enter the number of rows (m):"
    read *, m
    print *, "Enter the number of columns (n):"
    read *, n

    ! Allocate the matrix
    allocate(matrix(m, n))

    ! Read matrix elements
    do i = 1, m
        print *, "Enter elements of row ", i
        do j = 1, n
            read *, matrix(i, j)
        end do
    end do

    ! Print original matrix
    print *, "Original Matrix:"
    call printMatrix(matrix)

    ! Reverse rows
    call reverseRows(matrix)
    print *, "Matrix after reversing rows:"
    call printMatrix(matrix)

    ! Modify rows
    call modifyRows(matrix)
    print *, "Matrix after modifying rows:"
    call printMatrix(matrix)

    ! Deallocate the matrix
    deallocate(matrix)
end program matrix_operations

File: problema 19/c-languaje/problema_19.c
#include <stdio.h>

void printMatrix(double **matrix, int m, int n) {
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            printf("%.1f ", matrix[i][j]);
        }
        printf("\n");
    }
}

void reverseRows(double **matrix, int m, int n) {
    for (int i = 0; i < m / 2; i++) {
        for (int j = 0; j < n; j++) {
            double temp = matrix[i][j];
            matrix[i][j] = matrix[m - i - 1][j];
            matrix[m - i - 1][j] = temp;
        }
    }
}

void modifyRows(double **matrix, int m, int n) {
    double **tempMatrix = (double **)malloc(m * sizeof(double *));
    for (int i = 0; i < m; i++) {
        tempMatrix[i] = (double *)malloc(n * sizeof(double));
        for (int j = 0; j < n; j++) {
            tempMatrix[i][j] = m * matrix[i][j] + i * matrix[m - 1][j];
        }
    }
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            matrix[i][j] = tempMatrix[i][j];
        }
    }
    for (int i = 0; i < m; i++) {
        free(tempMatrix[i]);
    }
    free(tempMatrix);
}

int main() {
    int m, n;
    printf("Enter the number of rows (m): ");
    scanf("%d", &m);
    printf("Enter the number of columns (n): ");
    scanf("%d", &n);

    double **matrix = (double **)malloc(m * sizeof(double *));
    for (int i = 0; i < m; i++) {
        matrix[i] = (double *)malloc(n * sizeof(double));
        printf("Enter elements of row %d: ", i + 1);
        for (int j = 0; j < n; j++) {
            scanf("%lf", &matrix[i][j]);
        }
    }

    printf("Original Matrix:\n");
    printMatrix(matrix, m, n);

    reverseRows(matrix, m, n);
    printf("Matrix after reversing rows:\n");
    printMatrix(matrix, m, n);

    modifyRows(matrix, m, n);
    printf("Matrix after modifying rows:\n");
    printMatrix(matrix, m, n);

    for (int i = 0; i < m; i++) {
        free(matrix[i]);
    }
    free(matrix);

    return 0;
}